lib/util.js

```diff
@@ -748,27 +748,74 @@ var util = {
   /**
    * @api private
    */
-  addPromisesToRequests: function addPromisesToRequests(constructor, PromiseDependency) {
-    PromiseDependency = PromiseDependency || null;
-    if (!PromiseDependency && typeof Promise !== 'undefined') {
-      PromiseDependency = Promise;
-    }
-    if (typeof PromiseDependency !== 'function') {
-      delete constructor.prototype.promise;
-      return;
-    }
-    constructor.prototype.promise = function promise() {
-      var self = this;
-      return new PromiseDependency(function(resolve, reject) {
-        self.on('complete', function(resp) {
-          if (resp.error) {
-            reject(resp.error);
-          } else {
-            resolve(resp.data);
-          }
+  addPromises: function addPromises(constructors, PromiseDependency) {
+    if (!AWS) AWS = require('./core');
+    if (!AWS.config) require('./config');
+    if (PromiseDependency === undefined) PromiseDependency = AWS.config.getPromisesDependency();
+    if (typeof PromiseDependency !== 'function') var deletePromise = true;
+    if (!Array.isArray(constructors)) constructors = [constructors];
+
+    var promisifyMethod = function(methodName) {
+      return function promise() {
+        var self = this;
+        return new PromiseDependency(function(resolve, reject) {
+          self[methodName](function(err, data) {
+            if (err) {
+              reject(err);
+            } else {
+              resolve(data);
+            }
+          });
         });
-        self.runTo();
-      });
+      };
+    };
+
+    for (var ind = 0; ind < constructors.length; ind++) {
+      var constructor = constructors[ind];
+      switch (constructor.name) {
```

Comment:
Actually, what do you think about giving each class that should return a promise a static method that accepts a Promise constructor, then the class can control how it should promisify itself?

The pros to that approach would be the logic for adding promises would be controlled by each class, rather than defined in a long if/switch statement within a utility function. If the method to promisify a class was named the same for all classes, you can just check if the constructor has that method then call it, instead of maintaining a list of enums to check a class against. You could still make the `promisifyMethod` a utility method if that reduces code duplication.